File: sdk/nodejs/src/test/zbrt-connection.test.ts
import { describe, it, afterEach } from 'node:test';
import assert from 'node:assert/strict';
import net from 'node:net';
import { ZbrtFrame, KIND_EXECUTE, KIND_OUTPUT, KIND_EXIT, KIND_HEALTH, KIND_HEALTH_ACK, KIND_CANCEL, KIND_CANCEL_ACK, KIND_FS, KIND_FS_RESULT, decode } from '../zbrt-frame.js';
import * as codec from '../zbrt-codec.js';
import { ZbrtConnection } from '../zbrt-connection.js';
function frameBytes(kind: number, reqId: Buffer, payload: Buffer): Buffer {
  return new ZbrtFrame(kind, 0, reqId, payload).encode();
}

describe('ZbrtConnection (fake frame server)', () => {
  let server: net.Server;
  let port = 0;
  let cleanup: (() => void) | null = null;

  afterEach(() => {
    cleanup?.();
    cleanup = null;
  });

  function startServer(handler: (socket: net.Socket) => Promise<void>): Promise<number> {
    return new Promise((resolve) => {
      server = net.createServer((socket) => {
        void handler(socket);
      });
      server.listen(0, '127.0.0.1', () => {
        const addr = server.address() as net.AddressInfo;
        port = addr.port;
        // unref: the fake server must not keep the test process alive once
        // the suite's own sockets are closed.
        server.unref();
        cleanup = () => {
          server.close();
        };
        resolve(port);
      });
    });
  }

  it('execute collects output and returns exit', async () => {
    await startServer(async (socket) => {
      const frame = decode(await readFrame(socket));
      assert.equal(frame.kind, KIND_EXECUTE);
      const exec = codec.decodeExecute(frame.payload);
      assert.deepEqual(exec.argv, ['echo', 'hi']);
      const rid = frame.requestId;
      socket.write(frameBytes(KIND_OUTPUT, rid, codec.encodeOutput(0, Buffer.from('hello'))));
      socket.write(frameBytes(KIND_OUTPUT, rid, codec.encodeOutput(1, Buffer.from('err'))));
      socket.write(frameBytes(KIND_EXIT, rid, codec.encodeExit(0, null)));
    });

    const conn = new ZbrtConnection('127.0.0.1', port, 5000);
    await conn.ready();
    const result = await conn.execute(['echo', 'hi'], '/workspace', Buffer.alloc(0), 5000);
    assert.equal(result.code, 0);
    assert.deepEqual(result.stdout, Buffer.from('hello'));
    assert.deepEqual(result.stderr, Buffer.from('err'));
    assert.equal(result.timedOut, false);
    conn.close();
  });

  it('health roundtrips', async () => {
    await startServer(async (socket) => {
      const frame = decode(await readFrame(socket));
      assert.equal(frame.kind, KIND_HEALTH);
      socket.write(frameBytes(KIND_HEALTH_ACK, frame.requestId, codec.encodeHealth(true, 'ready')));
    });
    const conn = new ZbrtConnection('127.0.0.1', port, 5000);
    await conn.ready();
    const health = await conn.health();
    assert.equal(health.healthy, true);
    assert.equal(health.message, 'ready');
    conn.close();
  });

  it('cancel roundtrips with empty CancelAck', async () => {
    await startServer(async (socket) => {
      const frame = decode(await readFrame(socket));
      assert.equal(frame.kind, KIND_CANCEL);
      socket.write(frameBytes(KIND_CANCEL_ACK, frame.requestId, Buffer.alloc(0)));
    });
    const conn = new ZbrtConnection('127.0.0.1', port, 5000);
    await conn.ready();
    await conn.cancel('test', Buffer.alloc(16, 1));
    conn.close();
  });

  it('fs roundtrips payload', async () => {
    await startServer(async (socket) => {
      const frame = decode(await readFrame(socket));
      const fs = codec.decodeFs(frame.payload);
      assert.equal(fs.op, 4);
      assert.equal(fs.path, 'test.txt');
      socket.write(frameBytes(KIND_FS_RESULT, frame.requestId, Buffer.from(JSON.stringify({ data: [104, 105], truncated: false, total_bytes: 2 }))));
    });
    const conn = new ZbrtConnection('127.0.0.1', port, 5000);
    await conn.ready();
    const result = await conn.fs(4, 'test.txt', Buffer.from(JSON.stringify({ offset: null, max_bytes: 100 })));
    const parsed = JSON.parse(result.toString('utf8'));
    assert.deepEqual(parsed.data, [104, 105]);
    conn.close();
  });

  it('raises RemoteError on Error frame', async () => {
    await startServer(async (socket) => {
      const frame = decode(await readFrame(socket));
      socket.write(frameBytes(12, frame.requestId, codec.encodeError(1, 'guest error')));
    });
    const conn = new ZbrtConnection('127.0.0.1', port, 5000);
    await conn.ready();
    await assert.rejects(
      () => conn.execute(['echo'], '/workspace', Buffer.alloc(0), 5000),
      (error: Error) => error.message.includes('guest error'),
    );
    conn.close();
  });
});

/** Read exactly one frame (header + payload) from the socket. */
async function readFrame(socket: net.Socket): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    let buf = Buffer.alloc(0);
    const onData = (chunk: Buffer) => {
      buf = Buffer.concat([buf, chunk]);
      if (buf.length < 28) return;
      const len = buf.readUInt32BE(24);
      if (buf.length < 28 + len) return;
      socket.off('data', onData);
      socket.off('error', onError);
      resolve(buf.subarray(0, 28 + len));
    };
    const onError = (e: Error) => {
      socket.off('data', onData);
      socket.off('error', onError);
      reject(e);
    };
    socket.on('data', onData);
    socket.on('error', onError);
  });
}

File: sdk/java/tests/src/test/java/io/rfb/sdk/ZbrtConnectionTest.java
package io.rfb.sdk;

import io.rfb.sdk.internal.ZbrtCodec;
import io.rfb.sdk.internal.ZbrtConnection;
import io.rfb.sdk.internal.ZbrtFrame;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * ZBRT client against an in-process fake ZBRT server (PROTOCOL.md §3.4
 * session semantics: streamed Output → exactly one terminal, idempotent
 * Cancel → empty-payload CancelAck, Health → HealthAck, Error raise).
 */
class ZbrtConnectionTest {
    private FakeZbrtServer server;

    @AfterEach
    void tearDown() {
        if (server != null) {
            server.close();
        }
    }

    private ZbrtConnection open(FakeZbrtServer.ConnHandler handler) throws Exception {
        server = new FakeZbrtServer(handler);
        return new ZbrtConnection(
                io.rfb.sdk.internal.GuestNdjson.parseAddress(server.address()),
                java.time.Duration.ofSeconds(5));
    }

    @Test
    void helloReturnsServerNameAndCapabilities() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame hello = io.read();
            assertEquals(ZbrtFrame.KIND_HELLO, hello.kind());
            io.write(io.reply(hello, ZbrtFrame.KIND_HELLO_ACK,
                    ZbrtCodec.encodeHelloAck("rfb-zeroboot-guest", ZbrtConnection.V1_CAPABILITIES)));
        })) {
            ZbrtCodec.HelloAck ack = conn.hello("sdk-test");
            assertEquals("rfb-zeroboot-guest", ack.server);
            assertEquals(ZbrtConnection.V1_CAPABILITIES, ack.capabilities);
        }
    }

    @Test
    void executeCollectsOutputStreamsThenExit() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame exec = io.read();
            assertEquals(ZbrtFrame.KIND_EXECUTE, exec.kind());
            // encodeExecute: argc + texts + cwd flag + stdin + timeout — spot check
            io.writeFrames(
                    FakeZbrtServer.outputFrame(exec.requestId(), 0, "hi".getBytes(StandardCharsets.UTF_8)),
                    FakeZbrtServer.outputFrame(exec.requestId(), 1, "err".getBytes(StandardCharsets.UTF_8)),
                    FakeZbrtServer.exitFrame(exec.requestId(), 0));
        })) {
            ZbrtConnection.Exec exec = conn.execute(List.of("echo", "hi"), "/workspace",
                    "abc".getBytes(StandardCharsets.UTF_8), 1500);
            assertEquals(0, exec.code());
            assertEquals("hi", new String(exec.stdout(), StandardCharsets.UTF_8));
            assertEquals("err", new String(exec.stderr(), StandardCharsets.UTF_8));
            assertTrue(!exec.timedOut());
        }
    }

    @Test
    void errorFrameRaisesRemoteError() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame exec = io.read();
            io.write(FakeZbrtServer.errorFrame(exec.requestId(), 1, "argv is empty"));
        })) {
            RemoteError error = assertThrows(RemoteError.class,
                    () -> conn.execute(List.of(), null, new byte[0], 0));
            assertEquals(1, error.getCode());
            assertTrue(error.getMessage().contains("argv is empty"));
        }
    }

    @Test
    void secondExecuteAndDuplicateRequestRaise() throws Exception {
        AtomicInteger counter = new AtomicInteger(0);
        try (ZbrtConnection conn = open(io -> {
            while (true) {
                ZbrtFrame exec = io.read();
                int n = counter.incrementAndGet();
                if (n == 1) {
                    io.write(FakeZbrtServer.exitFrame(exec.requestId(), 0));
                } else if (n == 2) {
                    io.write(FakeZbrtServer.errorFrame(exec.requestId(), 1, "a turn is already active"));
                } else {
                    io.write(FakeZbrtServer.errorFrame(exec.requestId(), 1, "duplicate request id"));
                }
            }
        })) {
            assertEquals(0, conn.execute(List.of("x"), null, new byte[0], 0).code());
            RemoteError second = assertThrows(RemoteError.class,
                    () -> conn.execute(List.of("x"), null, new byte[0], 0));
            assertTrue(second.getMessage().contains("a turn is already active"));
            RemoteError third = assertThrows(RemoteError.class,
                    () -> conn.execute(List.of("x"), null, new byte[0], 0));
            assertTrue(third.getMessage().contains("duplicate request id"));
        }
    }

    @Test
    void cancelWithoutTargetExpectsEmptyCancelAck() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame cancel = io.read();
            assertEquals(ZbrtFrame.KIND_CANCEL, cancel.kind());
            ZbrtCodec.Cancel decoded = ZbrtCodec.decodeCancel(cancel.payload());
            assertEquals("user", decoded.reason());
            io.write(io.reply(cancel, ZbrtFrame.KIND_CANCEL_ACK, new byte[0]));
        })) {
            conn.cancel("user", null); // must return without raising
        }
    }

    @Test
    void cancelWithTargetRoundTripsTarget() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame cancel = io.read();
            ZbrtCodec.Cancel decoded = ZbrtCodec.decodeCancel(cancel.payload());
            byte[] target = decoded.target();
            if (target == null) {
                io.write(FakeZbrtServer.errorFrame(cancel.requestId(), 1, "missing target"));
                return;
            }
            io.write(io.reply(cancel, ZbrtFrame.KIND_CANCEL_ACK, new byte[0]));
        })) {
            conn.cancel(null, new byte[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16});
        }
    }

    @Test
    void healthReturnsReady() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame health = io.read();
            assertEquals(ZbrtFrame.KIND_HEALTH, health.kind());
            io.write(io.reply(health, ZbrtFrame.KIND_HEALTH_ACK, ZbrtCodec.encodeHealth(true, "ready")));
        })) {
            ZbrtCodec.Health health = conn.health();
            assertTrue(health.healthy());
            assertEquals("ready", health.message());
        }
    }

    @Test
    void fsReadAndWriteRoundTrip() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame fsRead = io.read();
            assertEquals(ZbrtFrame.KIND_FS, fsRead.kind());
            ZbrtCodec.Fs read = ZbrtCodec.decodeFs(fsRead.payload());
            assertEquals(4, read.op());
            assertEquals("notes.txt", read.path());
            io.write(io.reply(fsRead, ZbrtFrame.KIND_FS_RESULT,
                    "{\"data\":[104,105],\"truncated\":false,\"total_bytes\":2}"
                            .getBytes(StandardCharsets.UTF_8)));

            ZbrtFrame fsWrite = io.read();
            ZbrtCodec.Fs write = ZbrtCodec.decodeFs(fsWrite.payload());
            assertEquals(5, write.op());
            assertTrue(new String(write.data(), StandardCharsets.UTF_8).contains("\"data\":[104,105]"));
            io.write(io.reply(fsWrite, ZbrtFrame.KIND_FS_RESULT,
                    "{\"bytes_written\":2}".getBytes(StandardCharsets.UTF_8)));
        })) {
            byte[] readPayload = conn.fs(4, "notes.txt",
                    "{\"max_bytes\":51200}".getBytes(StandardCharsets.UTF_8));
            assertTrue(new String(readPayload, StandardCharsets.UTF_8).contains("\"data\":[104,105]"));
            byte[] writePayload = conn.fs(5, "notes.txt",
                    "{\"data\":[104,105],\"append\":false}".getBytes(StandardCharsets.UTF_8));
            assertTrue(new String(writePayload, StandardCharsets.UTF_8).contains("bytes_written"));
        }
    }

    @Test
    void executeReadStallRaisesTransportError() throws Exception {
        // Timeout is a transport-level failure: the client never cancels on
        // its own (mirrors the Rust baseline).
        server = new FakeZbrtServer(io -> {
            io.read(); // consume the Execute, then stay silent
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });
        try (ZbrtConnection conn = new ZbrtConnection(
                io.rfb.sdk.internal.GuestNdjson.parseAddress(server.address()),
                java.time.Duration.ofMillis(300))) {
            long start = System.nanoTime();
            assertThrows(TransportError.class,
                    () -> conn.execute(List.of("sleep"), null, new byte[0], 300));
            assertTrue((System.nanoTime() - start) < 3_000_000_000L, "timeout path must stay fast");
        }
    }

    @Test
    void requestIdMismatchRaisesDecodeError() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame exec = io.read();
            byte[] otherId = new byte[16];
            Arrays.fill(otherId, (byte) 0xAB);
            io.write(FakeZbrtServer.exitFrame(otherId, 0));
        })) {
            assertThrows(DecodeError.class, () -> conn.execute(List.of("x"), null, new byte[0], 0));
        }
    }

    @Test
    void streamSessionYieldsOutputEventsThenTerminalAndStopCancels() throws Exception {
        try (ZbrtConnection conn = open(io -> {
            ZbrtFrame exec = io.read();
            assertEquals(ZbrtFrame.KIND_EXECUTE, exec.kind());
            io.write(FakeZbrtServer.outputFrame(exec.requestId(), 0, "line".getBytes(StandardCharsets.UTF_8)));
            ZbrtFrame cancel = io.read();
            assertEquals(ZbrtFrame.KIND_CANCEL, cancel.kind());
            ZbrtCodec.Cancel decoded = ZbrtCodec.decodeCancel(cancel.payload());
            assertArrayEquals(exec.requestId(), decoded.target());
            io.write(io.reply(cancel, ZbrtFrame.KIND_CANCEL_ACK, new byte[0]));
            io.write(FakeZbrtServer.exitFrame(exec.requestId(), -1));
        })) {
            ZbrtConnection.ZbrtStreamSession session = conn.openStreamSession(
                    List.of("tail", "-f"), null, new byte[0], 0);
            ZbrtConnection.Event chunk = session.nextEvent();
            assertEquals(0, chunk.stream());
            assertEquals("line", new String(chunk.data(), StandardCharsets.UTF_8));
            session.stop(); // sends Cancel targeted at this request; expects CancelAck
            ZbrtConnection.Event exit = session.nextEvent();
            assertTrue(exit.isExit());
            assertEquals(Integer.valueOf(-1), exit.code());
        }
    }

    // ---- Sandbox facade over the ZBRT transport ------------------------------

    @Test
    void sandboxExecOverZbrtMatchesNdjsonShape() throws Exception {
        server = new FakeZbrtServer(io -> {
            ZbrtFrame exec = io.read();
            io.writeFrames(
                    FakeZbrtServer.outputFrame(exec.requestId(), 0, "hi".getBytes(StandardCharsets.UTF_8)),
                    FakeZbrtServer.exitFrame(exec.requestId(), 0));
        });
        RfbClient client = new RfbClient("http://127.0.0.1:1", "", 5.0);
        SandboxInfo info = new SandboxInfo();
        info.setId("sb-1");
        info.setGuestAddr(server.address());
        Sandbox sandbox = Sandbox.attach(client, info, RfbClient.TRANSPORT_ZBRT);
        ExecResult result = sandbox.exec(List.of("echo", "hi"), "/workspace", 60.0);
        assertEquals(Integer.valueOf(0), result.getExitCode());
        assertEquals("hi", result.stdoutText());
        assertEquals("", result.stderrText());
        assertEquals(Boolean.FALSE, Boolean.valueOf(result.isTimedOut()));
    }

    @Test
    void sandboxPingOverZbrtUsesHealth() throws Exception {
        server = new FakeZbrtServer(io -> {
            ZbrtFrame health = io.read();
            assertEquals(ZbrtFrame.KIND_HEALTH, health.kind());
            io.write(io.reply(health, ZbrtFrame.KIND_HEALTH_ACK, ZbrtCodec.encodeHealth(true, "ready")));
        });
        RfbClient client = new RfbClient("http://127.0.0.1:1", "", 5.0);
        SandboxInfo info = new SandboxInfo();
        info.setId("sb-1");
        info.setGuestAddr(server.address());
        Sandbox sandbox = Sandbox.attach(client, info, RfbClient.TRANSPORT_ZBRT);
        assertTrue(sandbox.ping());
    }

    @Test
    void execValidatesArgsAndCwdBeforeSending() throws Exception {
        server = new FakeZbrtServer(io -> io.read());
        RfbClient client = new RfbClient("http://127.0.0.1:1", "", 5.0);
        SandboxInfo info = new SandboxInfo();
        info.setId("sb-1");
        info.setGuestAddr(server.address());
        Sandbox zbrt = Sandbox.attach(client, info, RfbClient.TRANSPORT_ZBRT);
        Sandbox ndjson = Sandbox.attach(client, info, RfbClient.TRANSPORT_NDJSON);
        // fail closed on both transports: empty argv and bad cwd never reach the wire
        assertThrows(ValidationError.class, () -> zbrt.exec(List.of(), "/workspace", 60.0));
        assertThrows(ValidationError.class, () -> zbrt.exec(List.of("x"), "a/../b", 60.0));
        assertThrows(ValidationError.class, () -> ndjson.exec(List.of(), "/workspace", 60.0));
        assertThrows(ValidationError.class, () -> ndjson.exec(List.of("x"), "C:/tmp", 60.0));
    }

    @Test
    void evalOverZbrtRejectsBlankCodeBeforeConnect() {
        RfbClient client = new RfbClient("http://127.0.0.1:1", "", 5.0);
        SandboxInfo info = new SandboxInfo();
        info.setId("sb-1");
        info.setGuestAddr("127.0.0.1:1"); // nothing is listening; nothing may be sent
        Sandbox sandbox = Sandbox.attach(client, info, RfbClient.TRANSPORT_ZBRT);
        assertThrows(ValidationError.class, () -> sandbox.eval("   "));
    }
}
